src/schema.jl

```diff
@@ -237,6 +239,150 @@ function apply_schema(t::ConstantTerm, schema::Schema, Mod::Type)
+everything below the call to `unprotect` is treated as special formula syntax.
+
+A `Protected` context is created inside a [`FunctionTerm`](@ref) automatically,
+but can be manually created with a call to [`protect`](@ref).
+```
+"""
+struct Protected{Ctx} end
+Base.broadcastable(x::Protected) = Ref(x)
+"""
+    protect(term::T)
+
+Create a [`Protected`](@ref) context for interpreting `term` (and descendents) during
+`apply_schema`.
+
+Outside a [`@formula`](@ref), acts as a constructor for the singleton `Protected{T}`.
+
+# Example
+
+```jldoctest; setup = :(using Random; Random.seed!(1))
+julia> d = (y=rand(4), a=[1:4;], b=rand(4))
+
+julia> f = @formula(y ~ 1 + protect(a+b));
+
+julia> modelmatrix(f.rhs, d)
+4×2 Array{Float64,2}:
+ 1.0  1.48861
+ 1.0  2.21097
+ 1.0  3.95192
+ 1.0  4.9999
+
+julia> d.a .+ d.b
+4-element Array{Float64,1}:
+ 1.4886128300795012
+ 2.210968202158536
+ 3.951916339835734
+ 4.999904658898614
+```
+"""
+protect(ctx) = Protected{ctx}()
+# construct singletons to avoid method ambiguities using Type{<:Protected}
+
+function apply_schema(t::FunctionTerm, schema::Schema, Mod::Type)
+    args = apply_schema.(t.args, schema, protect(Mod))
+    FunctionTerm(t.f, args, t.exorig)
+end
+
+apply_schema(t::FunctionTerm, schema::Schema, Ctx::Protected) =
+    FunctionTerm(t.f, apply_schema.(t.args, schema, Ctx), t.exorig)
+apply_schema(t, schema::Schema, Ctx::Protected) = t
+
+function apply_schema(t::FunctionTerm{typeof(protect)}, schema::Schema, Ctx::Type)
+    tt = only(t.args)
+    apply_schema(tt, schema, protect(Ctx))
+end
+
+# protect in Protected context is a no-op
+function apply_schema(t::FunctionTerm{typeof(protect)}, schema::Schema, Ctx::Protected)
+    tt = only(t.args)
+    apply_schema(tt, schema, Ctx)
+end
+
+"""
+    unprotect(term)
+    unprotect(::Protected{T})
+
+Inside a [`@formula`], removes [`Protected`](@ref) status for the argument
+term(s).  This allows the usual special [`@formula`](@ref) interpretation of
+calls to `+`, `&`, `*`, and `~` to be restored inside an otherwise
+[`Protected`](@ref) context.
+
+When called outside a `@formula`, unwraps `Protected{T}` to `T`.
+
+# Example
+
+```jldoctest; setup = :(using Random; Random.seed!(1))
+julia> d = (y=rand(4), a=[1.:4;], b=rand(4));
+
+julia> f = @formula(y ~ 1 - unprotect(a&b));
+
+julia> modelmatrix(f, d)
+4×1 Array{Float64,2}:
+  0.5113871699204988
+  0.5780635956829281
+ -1.855749019507202
+ -2.9996186355944543
+
+julia> 1 .- d.a .* d.b
+4×1 Array{Float64,2}:
+  0.5113871699204988
+  0.5780635956829281
+ -1.855749019507202
+ -2.9996186355944543
+```
+"""
+unprotect(::Protected{Ctx}) where Ctx = Ctx
+
+function apply_schema(t::FunctionTerm{typeof(unprotect)}, schema::Schema, Ctx::Protected)
+    tt = only(t.args)
+    apply_schema(tt, schema, unprotect(Ctx))
+end
+
+# un-protecting special syntax: &, +, and *
+# 
+# if these occur as children of a non-special call, they are protected by
```

Comment:
Here "special" seems to be used to mean "normal Julia syntax." :-)

Reply:
Nope, still means `+`, `&` etc.  The situation this is describing is something like

```
protect(unprotect(a * b))
```

or

```
log(unprotect(a & b))
# -> FunctionTerm(log, [InteractionTerm(Term(:a), Term(:b))], ...)
```

(since any non-special call implicitly protects it's children).  So in this case, teh call to `&` occurs as a child to a call to `log`, but in a non-protected context.